apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/MSALoginActivity.java

```diff
@@ -0,0 +1,396 @@
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * HTTP Client to get tokens.
+     */
+    private static HttpClient sHttpClient;
+
+    /**
+     * Refresh token if logged in.
+     */
+    private String mRefreshToken;
+
+    /**
+     * Refresh token scope.
+     */
+    private String mRefreshTokenScope;
+
+    /**
+     * Web view.
+     */
+    private WebView mWebView;
+
+    /**
+     * Authentication provider type.
+     */
+    private AuthenticationProvider.Type mAuthType;
+
+    @SuppressLint("SetJavaScriptEnabled")
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+
+        /* Init UI. */
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.msa_login);
+
+        Serializable rawAuthType = getIntent().getSerializableExtra(AuthenticationProvider.Type.class.getName());
+        mAuthType = (AuthenticationProvider.Type) rawAuthType;
+
+        /* Init API client only once. */
+        if (sHttpClient == null) {
+            sHttpClient = createHttpClient(this);
+        }
+
+        /* Configure web view. */
+        mWebView = findViewById(R.id.web_view);
+        mWebView.setWebChromeClient(new WebChromeClient());
+        mWebView.getSettings().setJavaScriptEnabled(true);
+
+        /* Show prompt or message that there will be no prompt to sign in. */
+        String cookie = CookieManager.getInstance().getCookie(AUTHORIZE_URL);
+        boolean compact = mAuthType == AuthenticationProvider.Type.MSA_COMPACT;
+        if (compact && cookie != null && cookie.contains("MSPPre")) {
+            mWebView.loadData(getString(R.string.signed_in_cookie), "text/plain", "UFT-8");
+        } else {
+            signIn(null);
+        }
+    }
+
+    @SuppressWarnings("unused")
+    public void signIn(View view) {
+        mWebView.setWebViewClient(new WebViewClient() {
+
+            @Override
+            public void onPageFinished(WebView view, String url) {
+                checkSignInCompletion(url);
+            }
+
+            @Override
+            public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
+                failSignIn(errorCode, description);
+            }
+
+            @Override
+            @TargetApi(Build.VERSION_CODES.M)
+            public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
+                failSignIn(error.getErrorCode(), error.getDescription());
+            }
+        });
+        boolean compactTicket = mAuthType == AuthenticationProvider.Type.MSA_COMPACT;
+        String responseType = compactTicket ? "token" : "code";
+        String[] scopes = compactTicket ? SCOPES_COMPACT : SCOPES_DELEGATE;
+        mWebView.loadUrl(AUTHORIZE_URL + REDIRECT_URI_PARAM + CLIENT_ID_PARAM + "&response_type=" + responseType +
+                "&scope=" + TextUtils.join("+", scopes));
```

Comment:
Move `&scope=` and `&response_type` to const since `&client_id=` is const.

Reply:
Thanks for the feedback but we can clean it up in another task. This is just copied from another place and we don't change any lines of code here.